ADR/ADR023-use-uid-in-urls.md

```diff
@@ -0,0 +1,43 @@
+# ADR23: Use UIDs in URLs to replace form-id
+
+Date: 08/01/24
+
+## Status
+Proposed
+
+## Context
+
+See this spike for details about the required changes https://trello.com/c/BruuEN5Z/1225-spike-change-ids-in-urls-of-forms-to-not-be-the-primary-key
+
+We need to switch away from putting database ids in our URLs. Currently, form urls look something like this:
+`https://submit.forms.service.gov.uk/form/1234/apply-for-an-apple/5678`
+
+Where `1234` is the form-id.
+
+We instead want to change our URLs so that they look either like this: 
+
+`https://submit.forms.service.gov.uk/form/51f5b6f7/apply-for-an-apple/5678`
+
+where the 8 char hex code `51f5b6f7` has replaced the form-id, or like this:
```

Comment:
How many different possibilities does an 8 char hex code give us?

Reply:
I think it would be 16^8 = 4,294,967,296

Reply:
Which for scale is about half the total number of Notify messages has ever sent

Reply:
Nice!

Reply:
Yes let's go with the 8-char hex as that's also easier to type

Reply:
Al and I were spiking it out, we can use `SecureRandom.uuid[0, 8]` where we're using the first 8 chars of a UUID. We still

Reply:
I think there is a small risk of generating a rude word, if we're worried about that we could look at something like [sqids](https://sqids.org/ruby) as suggested by @AP-Hunt.

Reply:
I had a thought earlier, how many forms do we generate in a year with our end to end tests as they currently work? 🤔 They currently run every time we do a release of forms-admin, forms-api or forms-runner, and create (and delete) one new form on each run.

If we don't care about collisions between different stages then that's equal to the number of PRs merged in a year for each of those repos. Since 2023-01-10, we've merged 569 PRs to forms-admin, 211 for forms-api, and 327 to forms-runner. That's 1107 releases a year. Sanity checking that, there are currently 758 forms in the production database, and at time of writing the highest id in the database is 1873, so it seems reasonable that the end to end tests have created over a 1000 forms since they started running. 

For niceness, let's say we're currently creating 1024 forms a year for our end to end tests. So assuming things continue in the same way, and we only care about production, I _think_ we're looking at consuming (1024/16^8)*100 percent , or 0.000002 % of the possibility space... so I think we're good 😁